kernel/zync.c

```diff
@@ -0,0 +1,318 @@
+			}
+			/* Weird returns are from old k_mutex */
+			k_spin_unlock(&zync->lock, key);
+			return zync->owner == NULL ? -EINVAL : -EPERM;
+		}
+		delta = MIN(mod, zync->rec_count);
+		zync->rec_count -= delta;
+		mod -= delta;
+	}
+#endif
+
+	K_ZYNC_ATOM_SET(mod_atom) {
+		val1 = modclamp(zync, old_atom.val + mod);
+		delta = val1 - old_atom.val;
+		new_atom.val = (mod_atom == reset_atom) ? 0 : val1;
+		new_atom.waiters = mod < 0 && delta != mod;
+	}
+
+	nowait = K_TIMEOUT_EQ(timeout, Z_TIMEOUT_NO_WAIT);
+	must_pend = mod < 0 && mod != delta;
+
+#ifdef CONFIG_ZYNC_RECURSIVE
+	if (must_pend && zync->cfg.recursive && _current == zync->owner) {
+		zync->rec_count += -(mod - delta);
+		mod = 0;
+		must_pend = false;
+	}
+#endif
+
+	if (delta > 0) {
+		prio_boost_reset(zync);
+	}
+
+	if (must_pend || delta > 0) {
+		Z_WAIT_Q_LAZY_INIT(&zync->waiters);
+	}
+
+	for (woken = 0; woken < delta; woken++) {
+		if (!z_sched_wake(&zync->waiters, 0, NULL)) {
+			break;
+		}
+		resched = true;
+	}
+
+	/* Old condvar API wants the count of threads woken as the return value */
+	if (delta >= 0 && mod_atom == reset_atom) {
+		delta = woken;
+	}
+
+	if (resched) {
+		K_ZYNC_ATOM_SET(mod_atom) {
+			new_atom.waiters = z_waitq_head(&zync->waiters) != NULL;
+		}
+	}
+
+	if (reset_atom != NULL) {
+		uint32_t newval = reset_atom == mod_atom ? 0 : 1;
+
+		if (IS_ENABLED(CONFIG_ZYNC_VALIDATE)) {
+			__ASSERT(newval == 0 || newval != reset_atom->val,
+				 "noop zync reset (mislocked condvar?)");
+		}
+		K_ZYNC_ATOM_SET(reset_atom) {
+			new_atom.val = newval;
+		}
+	}
+
+	if (must_pend && !nowait) {
+		prio_boost(zync, _current->base.prio);
+		pendret = z_pend_curr(&zync->lock, key, &zync->waiters, timeout);
+		key = k_spin_lock(&zync->lock);
+
+		mod -= delta;
+		K_ZYNC_ATOM_SET(mod_atom) {
+			new_atom.val = modclamp(zync, old_atom.val + mod);
+			delta2 = new_atom.val - old_atom.val;
+		}
+		delta += delta2;
+	} else if (must_pend && nowait) {
+		pendret = -EAGAIN;
+	}
+
+	if (delta < 0) {
+		set_owner(zync);
+	}
+
+	if (resched && zync->cfg.fair) {
+		z_reschedule(&zync->lock, key);
+	} else {
+		k_spin_unlock(&zync->lock, key);
+	}
+	return pendret < 0 ? pendret : abs(delta);
+}
+
+void z_impl_k_zync_reset(struct k_zync *zync, k_zync_atom_t *atom)
+{
+	k_spinlock_key_t key = k_spin_lock(&zync->lock);
+
+	atom->val = zync->cfg.atom_init;
+
+	while (z_waitq_head(&zync->waiters)) {
+		z_sched_wake(&zync->waiters, -EAGAIN, NULL);
+	}
+
+	k_spin_unlock(&zync->lock, key);
+}
+
+int32_t z_impl_z_pzync(struct z_zync_pair *mod_z,
+		       int32_t mod, k_timeout_t timeout)
+{
+	return k_zync(&mod_z->zync, &mod_z->atom, NULL, mod, timeout);
+}
+
+#ifdef CONFIG_USERSPACE
+
+/* We can be passed a valid k_zync in a syscall as either a bare
+ * object or as the zync field of a pair.  Sort of a struggle with the
+ * existing API.  Returns true if the validated object is a
+ * z_zync_pair.
+ */
+static bool vrfy_zync(void *p, bool init)
```

Comment:
As mentioned, this seemed clever when I wrote it but in practice ends up being a ton of code that IMHO isn't really doing much good.